clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -106,6 +112,10 @@ public Sender(KafkaClient client,
         this.client = client;
         this.accumulator = accumulator;
         this.metadata = metadata;
+        if (sendInOrder)
+            this.partitionsInFlight = new HashSet<>();
+        else
+            this.partitionsInFlight = null;
```

Comment:
Discussed offline with Becket - rework this patch to avoid the null checks elsewhere; i.e., make the accumulator more explicitly aware of the `sendInOrder` requirement